hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala

```diff
@@ -0,0 +1,277 @@
+      case GT(_, _) =>
+        if (flipped)
+          Interval(endpoint(v, 1), maximumValueByType(typ)) // key > value
+        else
+          Interval(minimumValueByType(typ), endpoint(v, -1)) // value > key
+      case GTEQ(_, _) =>
+        if (flipped)
+          Interval(endpoint(v, -1), maximumValueByType(typ)) // key >= value
+        else
+          Interval(minimumValueByType(typ), endpoint(v, 1)) // value >= key
+      case LT(_, _) =>
+        if (flipped)
+          Interval(minimumValueByType(typ), endpoint(v, -1)) // key < value
+        else
+          Interval(endpoint(v, 1), maximumValueByType(typ)) // value < key
+      case LTEQ(_, _) =>
+        if (flipped)
+          Interval(minimumValueByType(typ), endpoint(v, 1)) // key <= value
+        else
+          Interval(endpoint(v, -1), maximumValueByType(typ)) // value <= key
+    }
+  }
+
+  private val noData: (Set[IR], Array[Interval]) = (Set(), Array())
+
+  def processPredicates(p: KeyFilterPredicate, kType: Type): (Set[IR], Array[Interval]) = {
+    p match {
+      case KeyComparison(comp) =>
+        val (v, isFlipped) = if (IsConstant(comp.l)) (comp.l, false) else (comp.r, true)
+        Set[IR](comp) -> Array(openInterval(constValue(v), v.typ, comp.op, isFlipped))
+
+      case LiteralContains(comp: IR) =>
+        val ApplyIR(_, Seq(Literal(_, lit), _)) = comp
+        val intervals = (lit: @unchecked) match {
+          case x: IndexedSeq[_] => x.map(elt => Interval(endpoint(elt, -1), endpoint(elt, 1))).toArray
+          case x: Set[_] => x.map(elt => Interval(endpoint(elt, -1), endpoint(elt, 1))).toArray
+          case x: Map[_, _] => x.keys.map(elt => Interval(endpoint(elt, -1), endpoint(elt, 1))).toArray
+        }
+        Set[IR](comp) -> intervals
+
+      case IntervalContains(comp: IR) =>
+        val ApplySpecial(_, Seq(Literal(_, lit), _)) = comp
+        val intervals = lit match {
+          case null => Array[Interval]()
+          case i: Interval => Array(Interval(endpoint(i.left.point, i.left.sign), endpoint(i.right.point, i.right.sign)))
+        }
+        Set[IR](comp) -> intervals
+
+      case LocusContigComparison(comp) =>
+        val (v, Apply(_, Seq(locus))) = if (IsConstant(comp.l)) (comp.l, comp.r) else (comp.r, comp.l)
+        val interval = (constValue(v): @unchecked) match {
+          case s: String => getIntervalFromContig(s, locus.typ.asInstanceOf[TLocus].rg.asInstanceOf[ReferenceGenome])
+        }
+        Set[IR](comp) -> Array(interval)
+
+      case LocusPositionComparison(comp) =>
+        val (v, isFlipped) = if (IsConstant(comp.l)) (comp.l, false) else (comp.r, true)
+        val pos = constValue(v).asInstanceOf[Int]
+        val rg = kType.asInstanceOf[TLocus].rg.asInstanceOf[ReferenceGenome]
+        val intOrd = TInt32().ordering.intervalEndpointOrdering
+        val intervals = rg.contigs.indices
+          .flatMap { i =>
+            Interval.intersection(Array(openInterval(pos, TInt32(), comp.op, isFlipped)),
```

Comment:
You're only intersecting single intervals here. I think this could be made a bit more efficient using the `Interval.intersect` method (as opposed to the class object method).